course-bundle-spec.md

```diff
@@ -0,0 +1,125 @@
+
+tags: [sample, life-changing, gcp]
+logo: ./gcp-intro-course-logo.png
+badge: ./gcp-intro-course-badge.png
+
+# Estimated time to take the course, *in days*
+duration: 1
+
+# Resources that may be referenced below in steps
+resources: ...
+
+# The important part of a course which lists all of the activities
+steps: ...
+
+```
+
+Note that all of the main chunks of localized content (title, description, objectives, audience, and prerequisites) are HTML content that may be displayed in various contexts. All of these chunks will be sanitized according to [instruction-html-spec.md](./instruction-html-spec.md).
+
+### Resources
+
+While heavyweight activities like labs and quizzes must be defined elsewhere in the library, we allow simpler resources to be specified directly in the course's `qwiklabs.yaml`. They can then be referenced in the `steps` just like labs and quizzes.
+
+For details on how to specify resources, see the [Resource Spec](./resource-spec.md).
+
+### Steps
+
+The meat of a course is an ordered list of steps defining what a learner needs to do to complete the course. A step consists of a set of one or more activity options, along with some metadata. Valid activity types are:
+
+* `lab`
+* `quiz`
+* `resource`
+
+`lab` and `quiz` will reference content defined elsewhere in the library, and `resource` will reference content defined elsewhere in the same `qwiklabs.yaml` file.
+
+The overall format should look like:
+
+```yml
+steps:
+  - activity_options:
+    - type: resource
+      content: intro-video
+    prompt:
+      locales:
+        en: If you've never used GCP, we recommend this overview video.
+    optional: true
+
+  - activity_options:
+    - type: lab
+      content: intro-to-gcp
+
+  - activity_options:
+    - type: resource
+      content: choosing-compute
+
+  - activity_options:
+    - type: lab
+      content: intro-to-appengine-python
+    - type: lab
+      content: intro-to-kubernetes-engine
+    - type: lab
+      content: intro-to-cloud-functions
+    prompt:
+      locales:
+        en: Learn more about one of these common compute resources.
+
+  - activity_options:
+    - type: quiz
+      content: compute-quiz
+```
+
+The order in which steps are listed defines the order they will be displayed. When a step has multiple options, the learner will be expected to do exactly one of the activities.
+
+The full specification for a step is as follows:
+
+attribute        | required | type        | notes
+---------------- | -------- | ----------- | -----------------------------------------
+activity_options | ✓        | array       | `activity_options` is an array of dictionaries with the format:
+-- type          | ✓        | enum        | One of `lab`, `quiz`, `resource`
+-- content       | ✓        | string      | Reference to the unique identifier for the activity. TODO: How to fully qualify `id` on labs and quizzes?
```

Comment:
It can be ironed out later, but if people have opinions on it, I would love to have those stated here

Reply:
I'd like to limit what characters are allowed in a lab id (perhaps `[a-z0-9-]`) and then allow "external" labs to be namespaced (maybe `namespace/lab-id`) to reference something outside of the repo or outside of the library.

Other than restricting the valid characters in an id, I'm fine punting on the specifics of namespacing until later. It will probably be fairly abstract until Alexandria gets closer to completion anyway.